vmconnect mode: support all security modes when used in a Hyper-V environment

gpotter2 opened the pull request:
This PR adds support for security protocols other than plain RDP when using xrdp inside a Hyper-V VM. In this specific case, SSL-related security features (CredSSP, etc.) are handled by Hyper-V, so xrdp may advertise for NLA support (like Windows does by default) and other advanced security features (restricted admin, etc.) without actually handling them itself.

![image](https://github.com/user-attachments/assets/984df02d-8e8e-495f-80db-2fc60162d473)
(source: me)

This adds a `vmconnect` flag in the configuration that turns this behavior on. This was tested on Win11 24H2 using vmconnect.exe.

metalefty commented:
Awesome!

matt335672 commented:
Agreed!

@gpotter2 - I'll need to review and test this, but you've caught me at a fairly busy period for the next couple of weeks IRL. Please don't take my silence for a lack of enthusiasm.

I haven't read the spec on this yet, so this might be a silly question, but is this something that only works over the vsock transport? If so, we may be able to make some further improvements.

gpotter2 commented:
Hi, thanks for the enthusiasm.

Just to tamper expectations, I didn't do the hard task of adding CredSSP support. This only works when running behind the Hyper-V "proxy" as the host intercepts CredSSP and handles it itself, so we can advertise it even though we don't implement it yet.

> but is this something that only works over the vsock transport

Yes. This will only work when binding on vsock, when accessed through the port 2179 of the Hyper-V parent host, which is the scenario in which vmconnect is used. (or just the "Enhanced Mode" of Hyper-V in general).

vmconnect authenticates against the Hyper-V host, and not its inner VMs, which explains why this trickery exists.

I made this PR because in a hardened environment, the current requirement to use "RDP" security when running inside a VM listening on vsock cannot work.

gpotter2 commented on libxrdp/xrdp_sec.c:
```diff
@@ -2252,7 +2252,11 @@ xrdp_sec_incoming(struct xrdp_sec *self)
-    if (iso->selectedProtocol > PROTOCOL_RDP)
+    if (self->rdp_layer->client_info.vmconnect)
+    {
+        /* Security handled by host: do nothing. */
```
The important part is this, we just skip entirely the security part.

matt335672 replied:
We can probably make this dependent on running over a vsock. What do you think?

gpotter2 replied:
Technically you can also connect to xrdp using the undocumented `mstsc.exe /v:<VM-ID> /hvsocketserviceid:00000D3D-FACB-11E6-BD58-64006A7986D3` (in this case 0xD3D=3389 as a Hyper-V port), assuming you have a properly built .rdp file. In this case, it opens a AF_HYPERV directly without going through vmms.exe. To support this use case I'd leave it a separate flag.

Edit: I also think that there might be a use case somewhere of people using vsock binding inside Qemu, or another Hypervisor, which is also a good reason.

matt335672 commented:
I understand about the difficulty of adding CredSSP support!

Speak soon.

matt335672 commented:
@gpotter - apologies for introdusing a conflict into your PR. This happened when I merged #3455

Regarding the conversation above about the vsock transport, I think there's a slight misunderstanding. At the moment we support multiple connections to the same VM, and I believe your PR could break this - see #2077. This can be avoided by only supporting vmconnect on vsock sockets.

You might be able to help me with my test rig here - I'm somewhat new to this vsock stuff. I've set up a Windows 11 VM running a nested Ubuntu VM for testing. I've got your code running with `--enable-vsock`, but I can't get the enhanced session mode working. The mode is enabled in the Hyper-V settings, and I've got `port=vsock://-1:3389` in my xrdp.ini. When I use vmconnect.exe however, I'm seeing no connection attempt in the xrdp log. Any ideas what I might be missing?

matt335672 commented:
Nevermind the query above - I found I needed the following tickbox set on the VM settings to get it working:-

![image](https://github.com/user-attachments/assets/e24c67c8-f66b-49aa-9290-2effb1f9909e)

I'm out of time today. I hope to do some more testing tomorrow.

gpotter2 committed:
vmconnect mode: support all security modes when used in Hyper-V environment

gpotter2 commented:
Hi & sorry for the delay. I've rebased the PR.

> You might be able to help me with my test rig here

The minimum setup is:
1. Configure 'port' on `vsock://1:3389` (minimum listener)
2. Check that enhanced session mode is allowed by the Hyper-V host:
![image](https://github.com/user-attachments/assets/e21a3664-6df0-469c-963a-57044eacbdf1)
3. You may set `Set-VM -EnhancedSessionTransportType HvSocket <VMName>` if you want windows to directly try to use a Hyper-V socket. It's not necessary though, as it ends up doing so anyways after timeouting on VMBus.
4. Enable xrdp to start on boot and restart the VM. If your host is up for too long with no vsock available on 3389, Hyper-V will no longer try.

An outsider reader should also remember to compile with `--enable-vsock`.

You don't have to enable *any* of the guests services for this to work. (The guest service related to RDP cannot be disabled, so it doesn't appear in the menu.)

matt335672 commented:
Thanks for the info @gpotter2 

You're quite right about the guest services - I've disabled them again, and everything seems OK. At some stage I'll recreate the environment and figure out what was causing the problem.

matt335672 commented on libxrdp/xrdp_iso.c:
```diff
@@ -410,10 +453,17 @@ xrdp_iso_send_cc(struct xrdp_iso *self)
+                flags |= RESTRICTED_ADMIN_MODE_SUPPORTED;
+            }
+
             /* [MS-RDPBCGR] RDP_NEG_RSP */
```
Shouldn't we only set `RESTRICTED_ADMIN_MODE_SUPPORTED` if the client sets `RESTRICTED_ADMIN_MODE_REQUESTED` in the connection request PDU?

gpotter2 replied:
The client sends `RESTRICTED_ADMIN_MODE_REQUIRED` (emphasis on the "REQUIRED" part).

Whether the server supports RestrictedAdmin or not is only a matter of Windows version (Windows > 2008), at least according to [MS-RDPBCGR].

matt335672 replied:
I understand that, but at least for other values, it's common practice for the client (or server) to suggest the capability (e.g. fastpath support) and the server (or client) to respond with the chosen capability. Is that not the case here? Am I misunderstanding how restricted admin mode works?

gpotter2 replied:
I didn't really look into it that much honestly, but I don't really see why a server can't advertise for RestrictedAdmin support at this point.

If a client requires RestrictedAdmin and server answers here it doesn't, it fails. But you could imagine that a client doesn't "require" it, but does it opportunistically if available, as suggested by the "Supported" name. In any case, it doesn't affect functionality.

I should probably boot up a real Windows to see what it advertises when the client doesn't "require" it.

matt335672 commented on libxrdp/xrdp_iso.c:
```diff
@@ -124,8 +129,36 @@ xrdp_iso_negotiate_security(struct xrdp_iso *self)
+            LOG(LOG_LEVEL_ERROR, "Impossible case.");
+            rv = 1;
+        }
+    }
```
This bit of code can be merged with the bit below, as the `PROTOCOL_HYBRID` and `PROTOCOL_HYBRID_EX` blocks can only be triggered in vmconnect mode:-

```c
static int
xrdp_iso_negotiate_security(struct xrdp_iso *self)
{
    int rv = 0;
    struct xrdp_client_info *client_info = &(self->mcs_layer->sec_layer->rdp_layer->client_info);
    char protostr[64];
    int got_protocol = 0;
    int security_type_mask;

    /* Map the configuration from xrdp.ini to a mask of allowed
     * security types ([MS-RDPBCGR] 2.2.1.2.1)
     *
     * There's some oddness around PROTOCOL_RDP. This value is 0,
     * for compatibility reasons, and it's OK for the server to
     * suggest RDP as the fallback protocol if nothing else is
     * agreed on. Nowadays, classic RDP security should
     * not be used, if at all avoidable */

    /* At present we only support SSL and RDP security */
    if (client_info->security_layer == SECURITY_LAYER_RDP)
    {
        security_type_mask = PROTOCOL_RDP;
    }
    else
    {
        security_type_mask = PROTOCOL_SSL;
    }
    /* But VMConnect mode supports everything. */
    if (client_info->vmconnect)
    {
        security_type_mask |= PROTOCOL_HYBRID | PROTOCOL_HYBRID_EX;
    }

    /* Logically 'and' this value with the mask requested by the client, and
     * see what's left */
    protocol_mask_to_str(self->requestedProtocol, protostr, sizeof(protostr));
    LOG(LOG_LEVEL_INFO, "Client requested security types (RDP assumed) : %s",
        protostr);
    security_type_mask &= self->requestedProtocol;

    if (security_type_mask & PROTOCOL_HYBRID_EX)
    {
        /* Currently supported by VMConnect mode only */
        LOG(LOG_LEVEL_INFO, "Selected HYBRID_EX security");
        self->selectedProtocol = PROTOCOL_HYBRID_EX;
        got_protocol = 1;
    }
    else if (security_type_mask & PROTOCOL_HYBRID)
    {
        /* Currently supported by VMConnect mode only */
        LOG(LOG_LEVEL_INFO, "Selected HYBRID security");
        self->selectedProtocol = PROTOCOL_HYBRID;
        got_protocol = 1;
    }
    else if ((security_type_mask & PROTOCOL_SSL) != 0)
    {
        /* Can we do TLS? (basic check) */
        if (g_file_readable(client_info->certificate) &&
                g_file_readable(client_info->key_file))
        {
            LOG(LOG_LEVEL_INFO, "Selected TLS security");
            self->selectedProtocol = PROTOCOL_SSL;
            got_protocol = 1;
        }
        else
        {
            LOG(LOG_LEVEL_WARNING, "Cannot accept TLS connections because "
                "certificate or private key file is not readable. "
                "certificate file: [%s], private key file: [%s]",
                client_info->certificate, client_info->key_file);

            /* If we're configured to ONLY use TLS, this is a problem.
             * If not, we can fall back to RDP */
            if (client_info->security_layer == SECURITY_LAYER_TLS)
            {
                LOG(LOG_LEVEL_ERROR,
                    "Server requires TLS (security_layer=tls)");
                self->failureCode = SSL_CERT_NOT_ON_SERVER;
                rv = 1;
            }
        }
    }
    else if (client_info->security_layer == SECURITY_LAYER_TLS)
    {
        /* We don't have a match on TLS, but we'll accept nothing less */
        LOG(LOG_LEVEL_ERROR, "Server requires TLS (security_layer=tls)");
        self->failureCode = SSL_REQUIRED_BY_SERVER;
        rv = 1;
    }

    /* If we haven't got a match so far, and we haven't got a fail,
     * try RDP */
    if (!got_protocol && !rv)
    {
        if (g_fips_mode_enabled())
        {
            /* This is a FIPS-mode machine, and we don't support classic RDP
             * encryption */
            LOG(LOG_LEVEL_ERROR,
                "Server in FIPS mode requires TLS for security");
            self->failureCode = SSL_REQUIRED_BY_SERVER;
            rv = 1; /* error */
        }
        else
        {
            self->selectedProtocol = PROTOCOL_RDP;
            LOG(LOG_LEVEL_INFO, "Selected classic RDP security");
            LOG(LOG_LEVEL_WARNING, "Classic RDP security is not secure -"
                " please configure TLS on the client and server");
            got_protocol = 1;
        }
    }

    return rv;
}
```

gpotter2 replied:
Will do !

gpotter2 replied:
A minor difference though: in vmconnect mode we aren't using the TLS certificates, so if TLS is negotiated we should skip the check.

matt335672 replied:
Interesting point - personally I'd rather leave the check in so the user doesn't get surprised when they also open up a TCP socket for comms, plus it keeps the code simpler. I can see the argument either way though.

matt335672 reviewed (commented):
A couple of comments on the code.

matt335672 commented:
@gpotter2 - thanks again for raising this.

A couple of questions for you:-
1) Are you aware of any available MS documentation on the interface between VMMS and the RDP service? I've had a trawl through the RDP stuff and failed to find anything in particular on this interface.
2) As expected, setting vmconnect mode breaks the functionality used in #2077. As a workaround, we can simply not allow vmconnect to work on anything other than a vsock. That's pretty straightforward to do, and I can pick that up if you like.

gpotter2 commented:
> Are you aware of any available MS documentation on the interface between VMMS and the RDP service?

AFAIK, there is none. It would also be slightly different on Windows compared to Linux, considering Windows uses the VMBus directly instead of a higher-level Hyper-V socket. The behavior described in this PR was understood by sniffing what comes out of the vsock, and comparing it to what comes in vmms.... although I should admit that it took me more time than necessary.

> As a workaround, we can simply not allow vmconnect to work on anything other than a vsock.

Sounds good. I can try to fix the other points if you'd like, although this'll have to wait tomorrow.

gpotter2 committed:
Apply suggestions

Co-Authored-By: matt335672 <30179339+matt335672@users.noreply.github.com>

gpotter2 commented:
I think I've addressed the code comments.

> we can simply not allow vmconnect to work on anything other than a vsock. That's pretty straightforward to do, and I can pick that up if you like.

If you have some guidance on how to detect that the session is bound on a vsock from this point in code, it'll make things faster.

matt335672 commented:
@gpotter2 - I'll try to get back to you with some code before the weekend, but I'm running short of time. From Sat I'm away from a keyboard for 10 days or so - as much as anyone can be these days.

matt335672 committed:
Disable vmconnect mode for non-vsock connections

matt335672 commented:
That turned out to be a lot simpler than I expected!

@gpotter2 - I've taken the liberty of pushing a small commit to your branch. Take a look and see if you're happy with it. 

On my Windows 11 VM, I've got a nested Ubuntu Hyper-V VM . Both machines are bridged to my test network. Ubuntu VM has following config:-

```
[Globals]
port=vsock://-1:3389 tcp://:3389
vmconnect=true
...
```

I can connect to the xrdp service using Enhanced mode, or from remmina on another machine.

gpotter2 commented:
Great !

matt335672 committed:
Updated manpage

matt335672 commented:
Just updated a manpage too. I've also added a brief description of the `port` setting, which was for some reason missing.

@gpotter2 - if you're happy with the manpage updates, I think we can merge this. I've had a look at the RestrictedAdmin feature above on my test rig, and using `vmconnect.exe` I'm not seeing `RESTRICTED_ADMIN_MODE_REQUESTED`, but us advertising `RESTRICTED_ADMIN_MODE_SUPPORTED` doesn't seem to be causing any trouble.

gpotter2 commented on docs/man/xrdp.ini.5.in:
```diff
@@ -23,6 +23,25 @@ All options and values (except for file names and paths) are case insensitive, a
+
+.TP
+\fBvmconnect\fR=\fI[true|false]\fR
+(Hyper-V VMs only). Allows the user to skip some unnecessary security
+checks when a virtual machine running xrdp is hosted on Hyper-V,
+and the user connects to it via the vmms service on TCP port 2179.
```
I find the "unnecessary checks" part a bit confusing. The rest is good.
Here's a suggestion that just omits it:

```suggestion
(Hyper-V VMs only). Enables a wider support of security protocols when a virtual machine running xrdp is hosted on Hyper-V,
and the user connects to it via the vmms service on TCP port 2179.
```

matt335672 replied:
Agreed.

gpotter2 commented on xrdp/xrdp.ini.in:
```diff
@@ -27,6 +27,10 @@ port=3389
 ; prefer use vsock://<cid>:<port> above
 use_vsock=false
```
Btw, if this is not recommended, it could be removed from the .ini's default, while still being supported for backwards compatibility. I find it slightly confusing.

matt335672 replied:
Agreed. Parameter removed and warning generated if the user still uses it.

gpotter2 commented:
@matt335672 Great ! I have minor comments but feel free to merge anytime. Current code looks good to me.

matt335672 committed:
Address review comments

matt335672 commented:
@gpotter2 - thanks again for one of the more interesting PRs we've had in a while!

gpotter2 commented:
My pleasure ! Thanks for the help and responsiveness.
